vendorconsent/tcf2/pubrestrict.go

```diff
@@ -0,0 +1,74 @@
+package vendorconsent
+
+import (
+	"fmt"
+
+	"github.com/prebid/go-gdpr/bitutils"
+)
+
+// IAB spec does not specify a max vendorID for the publisher restrictions. This should be one bit short of the max possible.
+const assumedMaxVendorID uint16 = 32767
+
+func parsePubRestriction(metadata ConsentMetadata, startbit uint) (*pubRestrictions, uint, error) {
+	data := metadata.data
+	numRestrictions, err := bitutils.ParseUInt12(data, startbit)
+	if err != nil {
+		return nil, 0, fmt.Errorf("parsePubRestriction - error on parsing the number of restrictions: %s", err.Error())
+	}
+
+	// Parse out the "exceptions" here.
+	currentOffset := startbit + 12
+	restrictions := make(map[byte]pubRestriction, numRestrictions)
+	for j := uint16(0); j < numRestrictions; j++ {
+		restrictData, err := bitutils.ParseByte8(data, currentOffset)
+		if err != nil {
+			return nil, 0, fmt.Errorf("parsePubRestriction - error on parsing the restriction purpose/type: %s", err.Error())
+		}
+		currentOffset = currentOffset + 8
+		numEntries, err := bitutils.ParseUInt12(data, currentOffset)
+		if err != nil {
+			return nil, 0, fmt.Errorf("parsePubRestriction - error on parsing the number of vendor ranges: %s", err.Error())
+		}
+		currentOffset = currentOffset + 12
+		vendors := make([]rangeConsent, numEntries)
+		for i := uint16(0); i < numEntries; i++ {
+			thisConsent, bitsConsumed, err := parseRangeConsent(data, currentOffset, assumedMaxVendorID)
+			if err != nil {
+				return nil, 0, err
+			}
+			vendors[i] = thisConsent
+			currentOffset = currentOffset + bitsConsumed
+		}
+		restrictions[restrictData] = pubRestriction{
+			purposeID:    (restrictData & 0xfc) >> 2,
```

Comment:
What are these hex values for?

Reply:
Masking the bits I am interested in, the high 6 bits or the low 2 bits. I probably don't need the `0xfc` mask, as the low bits will be shoved out of the way. But seems safer and clearer (when I was in the middle of it) to do it this way.